src/image.c

```diff
@@ -283,53 +324,46 @@ void print_image_no_colors(const Image* const image, const int chars, FILE *f) {
 	#else
 	char line[image->width * MB_LEN_MAX + 1];
 	#endif
-	int curLinePos;
 	line[image->width * MB_LEN_MAX] = 0;
 #endif
 
-	for ( y=0; y < image->height; ++y ) {
+	for ( int y=0; y < image->height; ++y ) {
 
-#if ! ASCII
-		curLinePos = flipx? image->width * MB_LEN_MAX : 0;
-#endif
-		for ( x=0; x < image->width; ++x ) {
+		int curLinePos = 0;
+		for ( int x=0; x < image->width; ++x ) {
 
-			const float lum = image->pixel[x + (flipy? image->height - y - 1 : y) * image->width];
-			const float opacity = image->alpha[x + (flipy? image->height - y - 1 : y) * image->width];
+			const int pixel_index = get_pixel_index(image, x, y);
+			const float lum = image->pixel[pixel_index];
+			const float opacity = image->alpha[pixel_index];
 			const int pos = ROUND((float)chars * lum);
 
 			int i = invert? pos : chars - pos;
 			i = ROUND((float)i * opacity);
+
+			char* char_dest = &line[curLinePos];
 #if ASCII
-			line[flipx? image->width - x - 1 : x] = ascii_palette[i];
+			char* char_start = &ascii_palette[i];
+			size_t char_len = 1;
 #else
-			int paletteI = ascii_palette_indizes[i];
-			if ( flipx )
-				curLinePos -= ascii_palette_lengths[i];
-			line[curLinePos++] = ascii_palette[paletteI];
-			// Add as many bytes as the char's length
-			for ( size_t j = 1; j < ascii_palette_lengths[i]; j++ ) {
-				line[curLinePos++] = ascii_palette[++paletteI];
-			}
-			if ( flipx )
-				curLinePos -= ascii_palette_lengths[i];
+			char* char_start = &ascii_palette[ascii_palette_indizes[i]];
+			size_t char_len = ascii_palette_lengths[i];
 #endif
-		}
-#if ASCII
-		fprintf(f, !use_border? "%s\n" : "|%s|\n", line);
-#else
-		if ( !flipx ) {
+			const vec2 gradient = get_image_gradient(image, x, y);
+			if( i > 1 && magnitude(gradient) > edge_threshold ) {
+				float direction_scaled = direction(gradient) / M_PI * 4. + 5.5;
```

Comment:
Where does 5.5 come from?
I guess this is 360° - 45° = 2 * π - (π/4). Rewriting this using π or a comment would be nice.

Reply:
Yes I think so, there is a +4 there that is needed to make the `fmod` call always return a positive number too. I have rewritten this part and added a comment :)